### DISCUSSION TITLE
Nimble central role support

### DISCUSSION THREAD

**Discussion by Johnmcenroyy** [InfiniTimeOrg/InfiniTime#1527] [Ideas]
### Verification

- [X] I searched for similar feature request and found none was relevant.

### Pitch us your idea!

Pinetime support of external bluetooth sensors 

### Description

Hello
Thank you very much for such a great software.
It will be great if it possible to add Nimble central role to connect to external sensors
such as Polar H10, Polar Stride Sensor, Zwift RunPod, external bluetooth gps (android can be as external gps
device over bluetooth - Bluetooth GPS Output, Share GPS) and so on.

Thanks

**Comment by Johnmcenroyy** [comment]
Just investigated a little more.

There is Adafruit Feather nRF52 Bluefruit: 
  https://www.adafruit.com/product/3406
And Adafruit Bluefruit52Lib for working with this module with Central mode example:
  https://github.com/adafruit/Adafruit_nRF52_Arduino/tree/master/libraries/Bluefruit52Lib/examples/Central
Also example of using Bluefruit52Lib for receive data from Polar heart rate sensor (H7,H10):
  https://github.com/MakerLabCRI/PolarBleArduino/blob/master/nRF/nRFPolarH10/nRFPolarH10.ino


Also there is Nordic example for using BLE Central - BLE Heart Rate Collector Example
  https://infocenter.nordicsemi.com/index.jsp?topic=%2Fsdk_nrf5_v17.1.0%2Fble_sdk_app_hrc.html&cp=8_1_4_2_0_1
Main source file for this example from SDK (<InstallFolder>\examples\ble_central\ble_app_hrs_c)
  https://termbin.com/fcfq

Maybe it can be integrated in InfiniTime firmware.

Update:
Sorry, just realized that all examples are for SoftDevice stack and NOT NimBLE

**Comment by Johnmcenroyy** [comment]
NimBLE Central example

[blecent](https://github.com/apache/mynewt-nimble/tree/master/apps/blecent): A basic central device with no user interface. This application scans for a peripheral that supports the alert notification service (ANS). Upon discovering such a peripheral, blecent connects and performs a characteristic read, characteristic write, and notification subscription.

**Comment by Riksu9000** [comment]
Can you explain what the purpose of connecting these devices to InfiniTime is?

**Comment by Johnmcenroyy** [comment]
> Can you explain what the purpose of connecting these devices to InfiniTime is?

Hi **Riksu9000**

I have just listed some sensors as example, the matter is to have BLE Central possibility to connect to external sensors.
It can be not just sensors but some bluetooth toys as PowerUP Smartphone Controlled Paper Airplanes and so on.
What concerns certain sensors, I am especially interested to connect from Pinetime to Polar H7 to get medical grade heart rate data, to get R-R intervals data to analyze stress, also there is a possibility to get aerobic threshold, anaerbic threshold with DFA Alpha 1 from HRV from chest heart rate.

So the main point is the possibility to connect to external sensors using NimBLE Central possibility and let the user decide which sensor and how to use it, let him write some additional apps for certain sensors.

Peter Gamma wrote also about this here:
https://forum.pine64.org/showthread.php?tid=11308

**Comment by Johnmcenroyy** [comment]
**Polar H7/H10 Chest Heart Rate Sensor**
  medical grade heart rate data
  hrv data (r-r intervals)  
    hrv give the possibility to get stress index, aerobic threshold,
    anaerobic threshold, heart problems etc.

no need software signal processing (heart rate and r-r intervals), 
sensor just give accurate heart data over bluetooth

**Polar Stride Footpod**
   speed
   distance
   leg cadence
   stride length

**Zwift Runpod Footpod**
   speed
   distance
   leg cadence

Footpods when calibrated can be accurate as gps and more.
Work in forest, indoor etc.

**PowerUP Smartphone Controlled Paper Airplanes**
   just toys

**Comment by nmichaels** [comment]
I'm interested in getting this done. It looks like NimBLE supports central concurrently with peripheral. I don't have any experience with that or this project, though, so if anyone feels like pointing me in a direction I'm happy to be led.

I looked earlier at adding ANT+ support, but it seems that the drivers require a [softdevice](https://www.thisisant.com/developer/components/nrf52832) and it looks like InfiniTime has grown beyond the world of proprietary protocol support. Oh well, ANT+ is going to die anyway, since it's never getting encryption support so it won't survive the EU's RED directive.

Anyway, if anyone's already taken any steps towards this, or has suggestions beyond reading `src/components/ble/NimbleController.cpp`, I'd love to hear about it.
